### ISSUE TITLE
ByteString.Char8 messes up the user input

### ISSUE THREAD

**Issue by gelisam** [gelisam/hawk#138]
I know we already rejected pull request #121, which replaced all our uses of ByteString.Char8 with Text, and we have good reasons for that. However, we haven't made any progress towards supporting multiple different input formats, and until we do, we should at least make sure not to mangle the input.

    $ echo 'hé!' | hawk -m 'id'
    h├®!

My understanding is that we prefer ByteString to Text because we want to support workflows which manipulate binary data such as images. That's a noble goal, but only if ByteString is general enough to represent both binary data and text. Perhaps there is a generic ByteString which represents bytes without losing any information, and also a more specific ByteString.Char8 which does lose information? If so, we should switch from ByteString.Char8 to ByteString.

More information about ByteString.Char8: http://a-dimit.blogspot.ca/2012/04/strings-in-haskell.html

**Comment by GregorySchwartz** [comment]
Could the solution just be a flag? --binary, otherwise Text?

**Comment by gelisam** [comment]
Yes, that's a much better idea, let's do that instead :)

**Comment by GregorySchwartz** [comment]
Is it crazy unreasonable due to the number of ByteString calls? I haven't looked at the codebase yet, but if it's fairly straightforward I could try my hand at it.

**Comment by gelisam** [comment]
Thanks, please try! I don't think it would involve too many changes. Let's see:

* `runtime/System/Console/Hawk/Runtime.hs`: probably the trickiest bit, this is where you need to find a way to communicate the contents of the flag to the runtime system.
* `src/System/Console/Hawk/Runtime/HaskellExpr.hs`: make sure to change the types to match the previous file, the type checker cannot catch mismatches on HaskellExpr at compile time!
* `runtime/System/Console/Hawk/Representable.hs`: maybe a multiparameter type class? The delimiter should probably be `String`.
* `src/System/Console/Hawk/Args/Option.hs`: again, the delimiter should probably be `String`. That's also where you would add the `--binary` flag, obviously.
* `runtime/System/Console/Hawk/Args/Spec.hs`: same
* `src/System/Console/Hawk/UserExpr/CanonicalExpr.hs`: just add a type parameter
* `src/System/Console/Hawk/UserExpr/InputReadyExpr.hs`: same
* `src/System/Console/Hawk/UserExpr/OriginalExpr.hs`: same
* `src/System/Console/Hawk/UserPrelude/Defaults.hs`: should probably import both, for the user's convenience

**Comment by GregorySchwartz** [comment]
Yeah, grep returned an acceptable amount of "B.". We could have simple helper functions -- instead of `B.reverse` we have 

```
import qualified Data.ByteString.Char8 as B
import qualified Data.Text as T

class GenericString a where
    gReverse :: a -> a

instance GenericString B.ByteString where
    gReverse = B.reverse

instance GenericString T.Text where
    gReverse = T.reverse
```

Then all functions just have `f :: (GenericString a) => a -> a`. Of course, the only deciding factors of which type to use is at the very beginning with the flag. Do you know of a package that has this kind of typeclass already? Otherwise we just need to implement the function used here.

**Comment by gelisam** [comment]
It's not that simple, but this will certainly have to be part of the solution.

[`IsString`](http://hackage.haskell.org/package/base-4.9.0.0/docs/Data-String.html#t:IsString) from base will be useful to convert the String delimiters into the target string type, but it doesn't have `reverse` nor any other string operations.

[`mono-traversable`](https://hackage.haskell.org/package/mono-traversable) supports list-like operations for ByteString and Text.

I think a more string-specific package exists, but I don't remember what it's called nor what it provides over the above two. A quick google search did find a few interesting results though, [here](https://snoyberg.wordpress.com/2009/12/04/string-like/) is one by Michael Snoyman and [here](https://github.com/jwiegley/stringable) is one by John Wiegley.

**Comment by GregorySchwartz** [comment]
Eh, I just wrote the class anyway as it's not that many functions.

**Comment by gelisam** [comment]
Sure, that works too.
